core/src/main/java/io/parsingdata/metal/data/selection/ByToken.java

```diff
@@ -78,5 +78,18 @@ private static ParseValueList getAllValuesRecursive(final ParseGraph graph, fina
         if (head.isGraph()) { return tailResults.add(getAllValuesRecursive(head.asGraph(), definition)); }
         return tailResults;
     }
+    
+    public static ParseItemList getAllRoots(final ParseGraph graph, final Token definition) {
+        checkNotNull(graph, "graph");
+        checkNotNull(definition, "definition");
+        return getAllRootsRecursive(graph, definition);
+    }
 
+    private static ParseItemList getAllRootsRecursive(final ParseItem item, final Token definition) {
+        if (item.getDefinition() == definition) { return ParseItemList.create(item); }
```

Comment:
We've discussed this off-line, but I'll also write this here for future reference: We will now miss results from self-referencing `Token`s. In that case a graph can contain another graph with the same `Token` definition that can still be considered a 'root' graph.

And by self-referencing tokens I mean either:

*a*) anonymous `Token`s that parse `this` (e.g. linked lists), or;
*b*) `Token`s that eventually `sub()` to themselves (currently implemented using *a*).